File: codegenerator.py
"""Module containing the CodeGenerator (stage 4/5 of the pipeline)"""

import binarystring
from architecture import instructions, addressingmodes, registers
from csmtokens import TypedToken, tokentypes
from csmdefaults import defaults, symboltypes
from csmstructs import Memory, Operand, Scope, Symbol, Instruction, Register

class CodeGenerator:

    def __init__(self, global_scope: Scope, procedure_scopes: dict[str, Scope], 
                 config_table: dict[str, int]) -> None:
        """Constructor for a 'CodeGenerator' object"""
        
        self.__global_scope = global_scope
        self.__procedure_scopes = procedure_scopes
        self.__config_table: dict[str, int] = config_table

        # Only the values 0..n-1 need to be represented in binary
        self.__machine_operation_bits: int = binarystring.number_bits(instructions.NUMBER_INSTRUCTIONS - 1)
        self.__addressing_mode_bits: int = binarystring.number_bits(addressingmodes.NUMBER_MODES - 1)
        
        self.__number_registers: int = config_table[defaults.REGISTERS_CONFIG[0]] + registers.NUMBER_SPRS
        self.__number_memory_addresses: int = config_table[defaults.MEMORY_CONFIG[0]]

        self.__operand_bits: int = (binarystring.number_bits(self.__number_registers) if 
                                    self.__number_registers > self.__number_memory_addresses
                                    else binarystring.number_bits(self.__number_memory_addresses - 1)) + 1

        self.__offset: int = 0
        self.__index: int = 0

        # For caching purposes
        self.__total_bits: int = self.__machine_operation_bits + self.__addressing_mode_bits + 2*self.__operand_bits
        self.__number_gprs: int = self.__number_registers - registers.NUMBER_SPRS

    def __update_global_symbols(self) -> None:
        """update all global scope related symbols"""

        offset: int = self.__global_scope.number_instructions + self.__global_scope.number_variables

        for identifier, scope in self.__procedure_scopes.items():

            self.__global_scope.symbol_table[identifier].value = offset
            offset += scope.number_instructions + scope.number_variables

    def __resolve_operand(self, operand: Operand, scope: Scope) -> int:
        """Return the value an operand points to"""

        match operand.value.type:

            case tokentypes.LABEL:

                symbol: Symbol | None = scope.symbol_table.get(operand.value.value, None)

                if symbol is None:

                    # If the label is not in the local scope is is in the global scope
                    symbol = self.__global_scope.symbol_table[operand.value.value]
                
                return symbol.value

            case tokentypes.VALUE:

                return int(operand.value.value)

            case tokentypes.REGISTER:

                spr: Register | None = registers.get_spr(operand.value.value)

                if spr is None:

                    # GPR Zero cannot be accessible so the value must be wrapped between 1..n (inclusive)
                    return defaults.wrap_bounds(1, self.__number_gprs, int(operand.value.value)) * -1

                return (self.__number_gprs + spr.offset) * -1

    def __generate_machine_operation(self, instruction: Instruction, source_operand: Operand, destination_operand: Operand, 
                                     scope: Scope) -> str:
        """Generate the machine code (bits) that would represent a low level CPU instruction"""

        instruction_bits: str = binarystring.unsigned_int(instruction.opcode, self.__machine_operation_bits)


        addressing_mode_bits: str = binarystring.unsigned_int(
            #                                  An addressing mode will always be returned
            addressingmodes.get_addressing_mode(source_operand.addressing_mode.value).opcode, self.__addressing_mode_bits)
        
        source_operand_bits: str = binarystring.signed_int(self.__resolve_operand(source_operand, scope), self.__operand_bits)

        destination_operand_bits: str = binarystring.signed_int(self.__resolve_operand(destination_operand, scope), self.__operand_bits)

        return instruction_bits + addressing_mode_bits + source_operand_bits + destination_operand_bits
    
    def __update_local_symbols(self, scope: Scope, memory: Memory) -> None:
        """Update any local symbols and prematurely place any variables into the memory pool"""

        self.__offset += scope.number_instructions

        # Update the local symbols
        for symbol in scope.symbol_table.values():

            match symbol.type:

                case symboltypes.BRANCH:

                    # Update address to be relative to its position in memory instead of its position in the scope
                    symbol.value += self.__index

                case symboltypes.VARIABLE:

                    # Place variables at the end of the instructions
                    memory.insert_binary(self.__offset, binarystring.signed_int(int(symbol.value), self.__total_bits))
                    symbol.value = self.__offset
                    self.__offset += 1

    def __generate_code(self, scope: Scope, memory: Memory) -> None:
        """Generate the code and variables for a given scope"""

        self.__update_local_symbols(scope, memory)

        # Generate the machine code for each instruction
        for index, token in enumerate(scope.tokens):

            if token.type == tokentypes.INSTRUCTION:

                #                           An instruction will always be returned
                instruction: Instruction = instructions.get_instruction(token.value)

                default_operand: Operand = Operand(
                    TypedToken(tokentypes.ADDRESSING_MODE, addressingmodes.REGISTER.symbol, -1, -1),
                    TypedToken(tokentypes.VALUE, defaults.OPERAND_VALUE, -1, -1)
                )

                source_operand: Operand = Operand(scope.tokens[index+1], scope.tokens[index+2]) \
                                          if instruction.operands > 0 \
                                          else default_operand

                destination_operand: Operand = Operand(scope.tokens[index+4], scope.tokens[index+5]) \
                                               if instruction.operands > 1 \
                                               else default_operand

                memory.insert_binary(self.__index, self.__generate_machine_operation(instruction, source_operand, destination_operand, scope))
                self.__index += 1

        self.__index = self.__offset

    def run(self) -> tuple[Memory, int, int, int]:
        """run the CodeGenerator
           Will return the memory and the bits used by each part of the machine code"""

        self.__update_global_symbols()
        
        memory: Memory = Memory(self.__number_registers, self.__total_bits, self.__operand_bits)

        self.__generate_code(self.__global_scope, memory)

        for scope in self.__procedure_scopes.values():

            self.__generate_code(scope, memory)

        return (memory, self.__machine_operation_bits, self.__addressing_mode_bits, self.__operand_bits)

File: chadsembler/src/pipeline/codegenerator.rs
use std::collections::HashMap;
use crate::csm::structs::{Scope, Symbol, Operand, Instruction, Memory};
use crate::csm::defaults::{sysdefaults, SymbolTypes};
use crate::csm::tokens::{TokenTypes, TypedToken};
use crate::csm::binarystring;
use crate::csm::architecture::{instructions, addressingmodes, registers};

/// Update all global scope related symbols
fn update_global_symbols<'a>(global_scope: &mut Scope<'a>, procedure_scopes: &mut HashMap<&'a String, Scope>) {

    let mut offset = global_scope.num_instructions + global_scope.num_variables;

    for (identifier, scope) in procedure_scopes {

        let symbol = global_scope.symbol_table.get_mut(identifier).unwrap();
        symbol.symbol_value = offset as isize;
        offset += scope.num_instructions + scope.num_variables;
    }
}

/// Return the value an operand points to
fn resolve_operand(operand: &Operand, scope_symbol_table: &HashMap<&String, Symbol>, global_symbol_table: &HashMap<&String, Symbol>, number_gprs: usize) -> isize {

    match operand.operand_value.token_type {

        TokenTypes::Label => {

            if let Some(s) = scope_symbol_table.get(&operand.operand_value.token_value) { 
                
                s.symbol_value 
            
            } else { 
                
                // If the symbol isn't in the local scope then it is in the global scope
                global_symbol_table[&operand.operand_value.token_value].symbol_value
            }
        },
        
        TokenTypes::Value => {

            operand.operand_value.token_value.parse().unwrap()
        },

        TokenTypes::Register => {

            if let Some(spr) = registers::SP_REGISTERS.get(&operand.operand_value.token_value) {
                
                (number_gprs + spr.offset) as isize * -1
                
            } else {
                
                // GPR Zero cannot be accessible so the value must be wrapped between 1..n (inclusive)
                sysdefaults::wrap_bound(1, (number_gprs) as isize, 
                                     operand.operand_value.token_value.parse::<isize>().unwrap())  * -1 
            }
        },

        _ => unreachable!()
    }
}

/// Generate the machine code (bits) that would represent a low level CPU instruction
fn generate_machine_operation(instruction: &Instruction, source_operand: &Operand, destination_operand: &Operand,
                              scope_symbol_table: &HashMap<&String, Symbol>, global_symbol_table: &HashMap<&String, Symbol>, number_gprs: usize,
                              machine_operation_bits: usize, addressing_mode_bits: usize, operand_bits: usize) -> String {

    let instruction_binary = binarystring::unsigned_int(instruction.opcode as isize, machine_operation_bits as isize);

    let addressing_mode_binary = binarystring::unsigned_int(addressingmodes::ADDRESSING_MODES[&source_operand.addressing_mode.token_value].opcode as isize, addressing_mode_bits as isize);
                            
    let source_operand_binary = binarystring::signed_int(resolve_operand(source_operand, scope_symbol_table, global_symbol_table, number_gprs), operand_bits as isize);
                            
    let destination_operand_binary = binarystring::signed_int(resolve_operand(destination_operand, scope_symbol_table, global_symbol_table, number_gprs), operand_bits as isize);
                            
    [instruction_binary, addressing_mode_binary, source_operand_binary, destination_operand_binary].concat()
}

/// Update any local symbols and prematurely place any variables into the memory pool
fn update_local_symbols(index: &mut usize, offset: &mut usize, scope: &mut Scope, memory: &mut Memory, total_bits: usize) {

    // The offset is where the variables should be placed in memory
    *offset += scope.num_instructions as usize;

    // Update the local symbols
    for symbol in scope.symbol_table.values_mut() {

        if symbol.symbol_type == SymbolTypes::Branch {

            //  Update address to be relative to its position in memory instead of its position in the scope
            symbol.symbol_value += *index as isize
        
        } else if symbol.symbol_type == SymbolTypes::Variable {

            //  Place variables at the end of the instructions
            memory.insert_binary(*offset as isize, binarystring::signed_int(symbol.symbol_value, total_bits as isize));
            symbol.symbol_value = *offset as isize;
            *offset += 1;
        }
    }
}

/// Generate the code for a given scope
fn generate_code(index: &mut usize, offset: &mut usize, scope_tokens: &Vec<&TypedToken>, scope_symbol_table: &HashMap<&String, Symbol>, 
                 global_symbol_table: &HashMap<&String, Symbol>, memory: &mut Memory,
                 number_gprs: usize, machine_operation_bits: usize, addressing_mode_bits: usize, operand_bits: usize) {

    let (mut current_source, mut current_destination);
    let default_operand = Operand {

        addressing_mode: &TypedToken {
            token_type: TokenTypes::AddressingMode,
            token_value: addressingmodes::REGISTER.symbol.to_owned(),
            row: 0,
            column: 0
        },

        operand_value: &TypedToken {
            token_type: TokenTypes::Value,
            token_value: sysdefaults::OPERAND_VALUE.to_owned(),
            row: 0,
            column: 0
        },
    };

    // Generate the machine code for each instruction
    for (i, token) in scope_tokens.iter().enumerate() {

        if token.token_type == TokenTypes::Instruction {

            let instruction = instructions::INSTRUCTION_SET[&token.token_value];

            let source_operand = if instruction.operands > 0 {

                current_source = Operand{ addressing_mode: scope_tokens[i+1], operand_value: scope_tokens[i+2] };
                &current_source
            
            } else {

                &default_operand
            };

            let destination_operand = if instruction.operands > 1 {

                current_destination = Operand{ addressing_mode: scope_tokens[i+4], operand_value: scope_tokens[i+5] };
                &current_destination
            
            } else {

                &default_operand
            };

            memory.insert_binary(*index as isize, generate_machine_operation(&instruction, source_operand, destination_operand, scope_symbol_table, global_symbol_table, number_gprs, machine_operation_bits, addressing_mode_bits, operand_bits));
            *index += 1;
        }
    }

    *index = *offset;
}

/// Run the CodeGenerator
/// Will return the memory and the bits used by each part of the machine code
pub fn run<'a>(global_scope: &mut Scope<'a>, procedure_scopes: &mut HashMap<&'a String, Scope>, config_table: &HashMap<String, usize>) -> (Memory, usize, usize, usize) {

    let machine_operation_bits = binarystring::number_bits(instructions::NUMBER_INSTRUCTIONS - 1);
    let addressing_mode_bits = binarystring::number_bits(addressingmodes::NUMBER_MODES - 1);

    let number_gprs = config_table[sysdefaults::REGISTERS_CONFIG.0];
    let number_registers = number_gprs + registers::NUMBER_SP_REGISTERS;

    let number_memory_addresses = config_table[sysdefaults::MEMORY_CONFIG.0] as usize;

    let operand_bits = if number_registers > number_memory_addresses
                            { binarystring::number_bits(number_registers) }
                            else { binarystring::number_bits(number_memory_addresses) } + 1;

    let total_bits = machine_operation_bits + addressing_mode_bits + 2*operand_bits;

    let (mut index, mut offset) = (0, 0);

    let mut memory = Memory::new(number_registers, total_bits, operand_bits);

    update_global_symbols(global_scope, procedure_scopes);

    update_local_symbols(&mut index, &mut offset, global_scope, &mut memory, total_bits);
    generate_code(&mut index, &mut offset, &global_scope.tokens, &global_scope.symbol_table, &global_scope.symbol_table, &mut memory, number_gprs, machine_operation_bits, addressing_mode_bits, operand_bits);

    for scope in procedure_scopes.values_mut() {

        update_local_symbols(&mut index, &mut offset, scope, &mut memory, total_bits);
        generate_code(&mut index, &mut offset, &scope.tokens, &scope.symbol_table, &global_scope.symbol_table, &mut memory, number_gprs, machine_operation_bits, addressing_mode_bits, operand_bits)
    }

    (memory, machine_operation_bits, addressing_mode_bits, operand_bits)
}
